train_decoder.py

```diff
@@ -0,0 +1,462 @@
+            train_images = generate_samples(decoder, dataloaders["train"], epoch, inference_device, step, n=n_sample_images, text_prepend="Train: ")
+            wandb.log({
+                "Test samples": test_images,
+                "Train samples": train_images
+            }, step=step)
+
+        print(f"=========== Starting Saving {epoch} ===========")
+        is_best = False
+        if save_best:
+            is_best = len(validation_losses) == 0 or average_loss < min(validation_losses)
+        validation_losses.append(average_loss)
+        save_trainer(base_path, trainer, epoch, step, validation_losses, local_only=not using_wandb, latest=True, best=is_best, checkpoint=False)
+
+    
+def initialize_training(config):
+    using_wandb = len(config["wandb"]["entity"]) > 0 and len(config["wandb"]["project"]) > 0
+    resuming = config["resume"]["do_resume"]
+    wandb_resume = config["resume"]["wandb_resume"]  # Whether to take over an old wandb run
+    resuming_from_source = None if not resuming else ("wandb" if config["resume"]["from_wandb"] else "file")
+
+    # Create the save path
+    base_path = "./models"
+    if not os.path.exists(base_path):
+        os.makedirs(base_path)
+    if "cuda" in config["train"]["device"]:
+        assert torch.cuda.is_available(), "CUDA is not available"
+    device = torch.device(config["train"]["device"])
+    torch.cuda.set_device(device)
+    all_shards = list(range(config["data"]["start_shard"], config["data"]["end_shard"] + 1))
+
+    dataloaders = create_dataloaders (
+        available_shards=all_shards,
+        img_preproc = config.get_preprocessing(),
+        train_prop = config["data"]["splits"]["train"],
+        val_prop = config["data"]["splits"]["val"],
+        test_prop = config["data"]["splits"]["test"],
+        **config["data"]
+    )
+
+    decoder = create_decoder(device, config["decoder"], config["unets"])
+
+    if using_wandb:
+        if wandb_resume:
+            run_id = config["resume"]["wandb_run_path"].split("/")[-1]
+            wandb.init(
+                project=config["wandb"]["project"],
+                entity=config["wandb"]["entity"],
+                id=run_id,
+                resume="must",
+                config=config.config,  # :P
+            )
+        else:
+            wandb.init(
+                project=config["wandb"]["project"],
+                entity=config["wandb"]["entity"],
+                config=config.config,
+            )
+
+    train(dataloaders, decoder,
+        inference_device=device,
+        resume_from=resuming_from_source,
+        resume_config=config["resume"],
+        using_wandb=using_wandb,
+        **config["train"],
+    )
+
+
+class TrainDecoderConfig:
+    def __init__(self, config):
+        defaults = {
+            "unets": [{ "dim": 16, "image_emb_dim": 768, "cond_dim": 64, "channels": 3, "dim_mults": [1, 2, 3, 4], "attn_dim_head": 32, "attn_heads": 16 }],
```

Comment:
hmm, couldn't this also be moved to a config?

Reply:
Yes, the defaults could be a separate file instead of hard coded. I generally prefer them to be hard coded so that people don't change the defaults instead of creating their own config, but if you prefer them to be separated I can make a separate file for them.

Reply:
yea, i think if you want to go config driven, just go all the way, otherwise we keep them as click argument defaults

Reply:
I removed the defaults to a `decoder_defaults.py` file. It's a python file instead of json because that allows for comments on the defaults, using more complex objects like a `REQUIRED` enum that makes it possible to have default values of None, and easier importing.